server/tso/global_allocator.go

```diff
@@ -84,8 +108,158 @@ func (gta *GlobalTSOAllocator) SetTSO(tso uint64) error {
 // GenerateTSO is used to generate a given number of TSOs.
 // Make sure you have initialized the TSO allocator before calling.
 func (gta *GlobalTSOAllocator) GenerateTSO(count uint32) (pdpb.Timestamp, error) {
-	// Todo: implement the synchronization algorithm for global TSO generation
-	return gta.timestampOracle.getTS(gta.leadership, count)
+	// To check if we have any dc-location configured in the cluster
+	dcLocationMap, err := gta.allocatorManager.GetClusterDCLocations()
+	if err != nil {
+		log.Error("get cluster dc-locations info failed", errs.ZapError(err))
+		return pdpb.Timestamp{}, err
+	}
+	// No dc-locations configured in the cluster
+	if len(dcLocationMap) == 0 {
+		return gta.timestampOracle.getTS(gta.leadership, count)
+	}
+	// Collect all allocator leaders' client URLs
+	allocatorLeaders, err := gta.allocatorManager.GetLocalAllocatorLeaders()
+	if err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	leaderURLs := make([]string, 0, len(allocatorLeaders))
+	for _, allocator := range allocatorLeaders {
+		for _, leaderURL := range allocator.GetMember().GetClientUrls() {
+			if slice.NoneOf(leaderURLs, func(i int) bool { return leaderURLs[i] == leaderURL }) {
+				leaderURLs = append(leaderURLs, leaderURL)
+			}
+		}
+	}
```

Comment:
I think we cannot calculate `leaderURLs` before entering `gta.syncMaxTS` because leader of local tso may drift to a PD that previously did not contain any tso leader so the PD will not appear in `leaderURLs`.